### PR TITLE
basic generics

### PR DESCRIPTION
**Pull request by dasnellings**
Now that Go 1.18 is out I have been playing around the generics. This PR removes the typed min/max functions in favor of a single generic min and max function in the numbers package. This is a fairly big change that would require everyone to update to 1.18 and introduces a new dependency to golang.org/x/exp as the generic packages are in googles experimental branch until generics get more field testing, however the expectation is these will move to the standard library in a future version of go. I think this would be worth a larger discussion before adding generics to gonomics so I am adding a do not merge tag for the time being.

### CODE REVIEW

**Comment by craiglowe** [comment]
This is exciting!  What is the down side?  I know you mentioned the dependency on exp and also needing everyone to be on 1.18 or greater.  Those seem acceptable to me (other people's thoughts?).  Any other read/potential downside that you see?

File: numbers/minMax.go
Comment (craiglowe) [inline numbers/minMax.go:9]: What do your benchmarks look like for this compared to just MinInt(a,b)?  I am wondering if turning the two ints into a slice slows things down?
Even if it does slow things down, maybe we could keep it but have two separate functions for min/max of two numbers, and min/max of any amount?

File: numbers/minMax.go
Comment (craiglowe) [inline numbers/minMax.go:14]: very minor, but since you are initializing to s[0], can we start the loop at 1 instead of 0?

**Review by craiglowe — COMMENTED** [review]
Lots of great stuff here.  I favor moving in this direction.  I added two small comments.

**Comment by craiglowe** [comment]
Does anyone have concerns about this?  Biggest thing I see would be making sure this doesn't break things for people when they go to run/compile main, but don't have >=1.18.  Can we have everyone upgrade their systems over the next couple days before we merge this so that once this gets merged, everyone is already on 1.18?

File: common/math_test.go
Comment (RavenYL) [inline common/math_test.go:18]: If we can go over what E is that would be helpful for me!

File: numbers/minMax_test.go
Comment (RavenYL) [inline numbers/minMax_test.go:165]: There might be other packages that need fixing. I remember using TripleMaxTrace in the align package, which compares scores in the 2x2 grid to decide which direction to proceed in the alignment matrix

**Review by RavenYL — COMMENTED** [review]
Although I'm always scared to go through updates and changes I agree that this is the direction we should go in

File: common/math_test.go
Comment (RavenYL) [reply common/math_test.go:18]: A Tour of Go's new section of generics has been helpful: https://go.dev/tour/generics/1

Comment (dasnellings) [reply common/math_test.go:18]: Yes! There is also a more in-depth tutorial that is pretty good https://go.dev/doc/tutorial/generics.

You probably figured it out from the Tour of Go, but the E was just a placeholder for the type name. In the blog I was looking at they used 'E'. In the Tour of Go they use 'T', but it could be anything.

File: numbers/minMax_test.go
Comment (dasnellings) [reply numbers/minMax_test.go:165]: I don't think the tripleMaxTrace function uses the TripleMax function so it shouldn't be affected.

**Review by craiglowe — APPROVED** [review]
Wonderful!  Looks fantastic!

**Review by RavenYL — APPROVED** [review]

**Comment by RavenYL** [comment]
I've upgraded to golang1.18 both locally and on DCC and am ready to merge when others are

**Pull request merged**
